add inherit_channel attribute for rjack-jetty

uchida opened the pull request:
This pull request add `inherit_channel` attribute to enable hot deploy, shutdown old process after new process is listening, with [`start_server`](https://metacpan.org/pod/start_server).

Originally [`setInheritChannel`](http://download.eclipse.org/jetty/9.3.6.v20151106/apidocs/org/eclipse/jetty/server/ServerConnector.html#setInheritChannel-boolean-) for `org.eclipse.jetty.server.Connector` is for `inetd` support, but this can use for hot deploy with `server_starter` >= 0.28

[Server::Starter meets Java](http://www.slideshare.net/tokuhirom/serverstarter-meets-java) describes `start_server` and how it works with Java web application.

A hot deploy sample is following:

1. start rjack-jetty sample code with `start_server`
  ```console
$ start_server --port=8080=0 --kill-old-delay=20 -- jruby server.rb
  ```
 with `--port=8080=0` `start_server` use port `8080` and file descriptor `0` for `inherit_channel`, and with `--kill-old-delay=20`, `start_server` wait 20 seconds to kill old process.
  `server.rb` code, borrowed from `rjack-jetty` sample, is following:
  ```ruby
require 'rjack-jetty'
require 'rjack-jetty/test-servlets'

include RJack

factory = Jetty::ServerFactory.new
factory.inherit_channel = true
factory.port = 8080
factory.set_context_servlets('/', '/*' => Jetty::TestServlets::SnoopServlet.new)

server = factory.create
server.start
server.join
  ```
  
2. hot deploy with `kill -HUP`. `start_server` launch new process and kill old process.
  ```console 
$ kill -HUP {start_server pid}
  ```

I also confirmed [fishwife](https://github.com/dekellum/fishwife/), rack server using `rjack-jetty`, work with this pull request by giving `-O inherited_channel` option.

dekellum commented:
Thanks for the request details.  In general support for this seems reasonable to include. I have two questions on specific:

1. What version(s) of Jetty support the connector.inherit_channel property?
1. Is this only usable for at most one channel and thus one connector, or will it work with multiple? If only the former, then the top-level ServerFactory property should probably not be included.

uchida commented:
> 1. What version(s) of Jetty support the connector.inherit_channel property?

From [jetty >= 9.0.0](https://github.com/eclipse/jetty.project/blob/jetty-9.0.0.M0/jetty-server/src/main/java/org/eclipse/jetty/server/ServerConnector.java#L167-L170), ServerConnector support has `setInheritedChannel` method.

For [jetty 8](http://download.eclipse.org/jetty/8.1.17.v20150415/apidocs/org/eclipse/jetty/server/nio/InheritedChannelConnector.html) and [jetty 7](http://download.eclipse.org/jetty/7.6.17.v20150415/apidocs/), there are no `setInheritedChannel` method, but `InheritedChannelConnector` is available, so using this instead of `SelectChannelConnector` will work.

> 2. Is this only usable for at most one channel and thus one connector, or will it work with multiple? If only the former, then the top-level ServerFactory property should probably not be included.

This will work multiple connector.

Dive into the detail, [jetty](https://github.com/eclipse/jetty.project/blob/jetty-9.3.6.v20151106/jetty-server/src/main/java/org/eclipse/jetty/server/ServerConnector.java#L311-L318) uses [System.inheritedChannel](http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b14/java/lang/System.java#System.inheritedChannel%28%29) and tracking this method in OpenJDK source codes, finally get [createChannel in sun.nio.ch.InheritedChannel](http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b14/sun/nio/ch/InheritedChannel.java#InheritedChannel.createChannel%28%29).
In this method, duplicating fd 0 with `dup(0)` and use when it is a socket.
And `start_server` with `--port=8080=0` option duplicate socket to fd 0 and pass fd0 to jetty.

Multiple `dup(0)` in [createChannel in sun.nio.ch.InheritedChannel](http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b14/sun/nio/ch/InheritedChannel.java#InheritedChannel.createChannel%28%29) will work for multiple connectors, so it will work for multiple connector.

uchida commented:
Sorry, above comment about multiple connectors seems inaccurate, because inheritechannel is wheather use stdin as socket or not, impossible to share stdin by connectors, so multiple connectors won't work.
I will remove top-level property later.

uchida commented:
removed top-level property and squashed commits.

dekellum commented:
With the squash I can only see your final change, not the delta, but `@inherit_channel` should no longer be required. In other words, to enable this, a users would just use the newer passed in opts syntax.

uchida commented:
Sorry to squash, it was a bad manner to review.
I recovered these commits from `git reflog` and re-push it see 9fc2036 and aa61484.

I added comments about `inherit_channel` option in aa61484.
And I realized connection `tcp://127.0.0.1?inherit_channel=true` doesn't work without changes, consult 1fde1800 for detail, so I also added codes to treat `opt[:inherit_channel]` as boolean in ea45d24.

uchida commented:
just to note, I replaced ea45d24 with 1393582, only a whitespace change.

uchida commented:
To make connection option style `tcp://127.0.0.1?inherit_channel=false` work as false, I added boolean conversion in 1393582.
But it seems complex, to keep simplicity of codes, omitting `tcp://127.0.0.1?inherit_channel=false` style could be an option.
@dekellum how do you feel about it? let me know.

uchida commented:
`inherit_channel` boolean conversion logic is changed in 7988d3d, when string/boolean true is given true and otherwise false.

dekellum commented:
Merged. Thanks for the cleanup and unit test and sorry for my delay, only due to holidays.  Given conventions of the project I would prefer to release this along with the upstream jetty 9.3.7 (i.e. 9.3.7.0) but I will consider a 9.3.6.1 if that doesn't happen soon.

uchida commented:
Thanks!
